File: profit_max.py
class Tree:

	def dfs_mod(self, start, visited, m, profit, max_p):
		visited[start] = True
		max_p = max(profit, max_p)

		neighbors_list = self.adj[start]
		for neighbor in neighbors_list:
			next = neighbor[0]
			t = neighbor[1]
			p = neighbor[2]

			if(m >= t and not visited[next]):
				temp = self.dfs_mod(next, visited, m - t, profit + p, max_p)
				max_p = max(max_p, temp)

		return max_p

	def __init__(self, n, adj):
		self.n = n
		self.adj = adj

n, m  = map(int, raw_input().strip().split())
structure = []
for i in xrange(n + 1):
	structure.append([])

for a0 in xrange(n - 1):
	u, v, t, p = map(int, raw_input().strip().split())
	structure[u].append([v, t, p])
	structure[v].append([u, t, p])

market = Tree(n, structure)

max_t = 0

for ind in range(1, len(structure)):
	visited = []
	for i in range(n + 1):
		visited.append(False)
	max_t = max(max_t, market.dfs_mod(ind, visited, m, 0, 0))

print max_t

File: profit_max.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class Tree {
public:
	unsigned int n;
	//A vector of all the nodes. Each node contains a vector representing it's neighbors(through exchanges). Each neighbor has a node number, ti and pi value.
	vector < vector < vector < unsigned long long int > > > adj;

	unsigned long long int dfs_mod(unsigned long long int start, vector<bool> &visited, unsigned long long int m, unsigned long long int profit, unsigned long long int max) {
		visited[start] = true;

		//printf("VISITING NODE %llu\n", start);
		if(profit > max){
			max = profit;
		}

		for(vector< vector <  unsigned long long int > >::iterator it = adj[start].begin(); it != adj[start].end(); it ++) {
			unsigned long long int next = (*it)[0];
			unsigned long long int t = (*it)[1];
			unsigned long long int p = (*it)[2];

			if(m >= t && !visited[next]) {
				unsigned long long int temp = dfs_mod(next, visited, m - t, profit + p, max);
				if (temp > max) {
					max = temp;
				}
			} 
		}

		return max;
	}

	Tree(vector < vector < vector < unsigned long long int > > > str, unsigned int num) {
		adj = str;
		n = num;
	}


};

int main(void){
	unsigned int n;
	unsigned long long int m;

	cin>>n>>m;
	unsigned long long int u, v, t, p;
	vector < vector < vector < unsigned long long int > > > structure(n + 1);	
	vector < unsigned long long int > data(3); 

	for(unsigned int i = 0; i < n - 1; i ++) {
		cin>>u>>v>>t>>p;
		data[0] = v;
		data[1] = t;
		data[2] = p;

		structure[u].push_back(data);
		data[0] = u;
		structure[v].push_back(data);

	}

	Tree market(structure, n);

	unsigned long long int max = 0;
	for(unsigned int i = 1; i < n + 1; i ++){
		vector<bool> visited_none(n + 1, false);
		unsigned long long int temp = market.dfs_mod(i, visited_none, m, 0, 0);
		if(temp > max){
			max = temp;
		}
	}

	cout<<max<<endl;

	return 0;
}